test/ThingsSpec.hs

```diff
@@ -1,6 +1,111 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
 module ThingsSpec where
 
-import Test.Hspec
+import Control.Lens
+import Control.Monad (liftM)
+import Control.Monad.Error.Class (MonadError, catchError, throwError)
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans.Except (ExceptT, throwE)
+import Crypto.JWT (StringOrURI, defaultJWTValidationSettings)
+import Crypto.Random.Types (MonadRandom, getRandomBytes)
+import Data.Aeson
+import qualified Data.Aeson.Lens as A
+import Data.Maybe (fromJust)
+import Data.Proxy
+import Data.String
+import Data.String.Conversions (cs)
+import Data.Text
+import Network.Wai
+import Network.Wai.Test (SRequest (..), simpleBody)
+import Servant.API
+import Servant.OAuth.Grants (OAuthGrantOpaqueAssertion (..), OpaqueToken (..))
+import Servant.OAuth.Grants.Facebook
+import Servant.OAuth.JWT
+import Servant.OAuth.ResourceServer
+import Servant.OAuth.TokenServer
+import Servant.OAuth.TokenServer.Types
+import Servant.Server
+import Test.Hspec hiding (pending)
+import Test.Hspec.Wai
+import Test.Hspec.Wai.Matcher
+
+------------------------------
+
+-- | generated with 'mkTestJWTSignSettings'
+testJWTSignSettings :: JWTSignSettings
+Just testJWTSignSettings =
+  decode "{\"jwtDuration\":5,\"jwtInitialClaims\":{},\"jwtSignKey\":{\"crv\":\"Ed25519\",\"d\":\"ZfSXWx4QCq4mQW_lPOXGvcqfEy6757Q2s9gWK2YbV88\",\"key_ops\":[\"sign\",\"verify\"],\"kid\":\"RHKw2tjb43P5mMab0m_xpYbNpAaiXROLdaOR8so4joo\",\"kty\":\"OKP\",\"use\":\"sig\",\"x\":\"Rm-3PqAInCgSjdlqWJz1RKADlIajHLa5So-uY4R95EU\"}}"
+
+testJWTSettings :: JWTSettings
+testJWTSettings =
+  JWTSettings
+    (SomeJWKResolver (jwtSignKey testJWTSignSettings))
+    (defaultJWTValidationSettings (== tokenPayload))
+
+tokenPayload :: IsString s => s
+tokenPayload = "..."
+
+------------------------------
+
+newtype AppM a = AppM {runAppM :: Handler a}
+  deriving newtype (Functor, Applicative, Monad, MonadIO)
+
+instance MonadRandom AppM where
+  getRandomBytes =
```

Comment:
As you already wrote (https://github.com/wireapp/servant-oauth-server/pull/5#issuecomment-1272797522), it would be great to figure out why this works.

Reply:
> As you already wrote ([#5 (comment)](https://github.com/wireapp/servant-oauth-server/pull/5#issuecomment-1272797522)), it would be great to figure out why this works.

when i do this:

```
~/src/hs-jose$ git diff
diff --git a/src/Crypto/JOSE/JWA/JWK.hs b/src/Crypto/JOSE/JWA/JWK.hs
index 57dd268..604952f 100644
--- a/src/Crypto/JOSE/JWA/JWK.hs
+++ b/src/Crypto/JOSE/JWA/JWK.hs
@@ -611,7 +611,7 @@ sign JWA.JWS.PS512 (RSAKeyMaterial k) = signPSS SHA512 k
 sign JWA.JWS.HS256 (OctKeyMaterial k) = signOct SHA256 k
 sign JWA.JWS.HS384 (OctKeyMaterial k) = signOct SHA384 k
 sign JWA.JWS.HS512 (OctKeyMaterial k) = signOct SHA512 k
-sign JWA.JWS.EdDSA (OKPKeyMaterial k) = signEdDSA k
+sign JWA.JWS.EdDSA (OKPKeyMaterial _k) = error "signEdDSA k"
 sign h k = \_ -> throwing _AlgorithmMismatch
   (show h <> " cannot be used with " <> showKeyType k <> " key")
```

and run the test suite, i get this:

```
  src/Crypto/JOSE/JWA/JWK.hs:614:42:
  1) Things, fetch token, success case
       uncaught exception: ErrorCall
       signEdDSA k
```

`signEdDSA` does not have a `MonadRandom` constraint (other cases in the `sign` function do, though, so there is no easy way to avoid the constraint in `servant-oauth-server`.

I checked if we *would* be using the right `MonadRandom` implementation:

```
~/src/hs-jose$ git diff
diff --git a/src/Crypto/JOSE/JWA/JWK.hs b/src/Crypto/JOSE/JWA/JWK.hs
index 57dd268..e9022b3 100644
--- a/src/Crypto/JOSE/JWA/JWK.hs
+++ b/src/Crypto/JOSE/JWA/JWK.hs
@@ -611,7 +611,7 @@ sign JWA.JWS.PS512 (RSAKeyMaterial k) = signPSS SHA512 k
 sign JWA.JWS.HS256 (OctKeyMaterial k) = signOct SHA256 k
 sign JWA.JWS.HS384 (OctKeyMaterial k) = signOct SHA384 k
 sign JWA.JWS.HS512 (OctKeyMaterial k) = signOct SHA512 k
-sign JWA.JWS.EdDSA (OKPKeyMaterial k) = signEdDSA k
+sign JWA.JWS.EdDSA (OKPKeyMaterial k) = \_ -> error . show . Types.Base64Octets =<< getRandomBytes 3
 sign h k = \_ -> throwing _AlgorithmMismatch
   (show h <> " cannot be used with " <> showKeyType k <> " key")
```

And it seems so:

```
  test/ThingsSpec.hs:77:5: 
  1) Things, fetch token, success case
       uncaught exception: ErrorCall
       Prelude.undefined
```

So everything appears to be in order!  Just think about whether we picked the coolest crypto primitives, or whether we need to put more thought into that.